rocksdb/src/main/java/com/yahoo/ycsb/db/rocksdb/RocksDBClient.java

```diff
@@ -0,0 +1,384 @@
+    }
+  }
+
+  private void saveColumnFamilyNames() throws IOException {
+    final Path file = rocksDbDir.resolve(COLUMN_FAMILY_NAMES_FILENAME);
+    try(final PrintWriter writer = new PrintWriter(Files.newBufferedWriter(file, UTF_8))) {
+      writer.println(new String(RocksDB.DEFAULT_COLUMN_FAMILY, UTF_8));
+      for(final String cfName : COLUMN_FAMILIES.keySet()) {
+        writer.println(cfName);
+      }
+    }
+  }
+
+  private List<String> loadColumnFamilyNames() throws IOException {
+    final List<String> cfNames = new ArrayList<>();
+    final Path file = rocksDbDir.resolve(COLUMN_FAMILY_NAMES_FILENAME);
+    if(Files.exists(file)) {
+      try (final LineNumberReader reader =
+               new LineNumberReader(Files.newBufferedReader(file, UTF_8))) {
+        String line = null;
+        while ((line = reader.readLine()) != null) {
+          cfNames.add(line);
+        }
+      }
+    }
+    return cfNames;
+  }
+
+  private Map<String, ByteIterator> deserializeValues(final byte[] values, final Set<String> fields,
+      final Map<String, ByteIterator> result) {
+    final ByteBuffer buf = ByteBuffer.allocate(4);
+
+    int offset = 0;
+    while(offset < values.length) {
+      buf.put(values, offset, 4);
+      buf.flip();
+      final int keyLen = buf.getInt();
+      buf.clear();
+      offset += 4;
+
+      final String key = new String(values, offset, keyLen);
+      offset += keyLen;
+
+      buf.put(values, offset, 4);
+      buf.flip();
+      final int valueLen = buf.getInt();
+      buf.clear();
+      offset += 4;
+
+      if(fields == null || fields.contains(key)) {
+        result.put(key, new ByteArrayByteIterator(values, offset, valueLen));
+      }
+
+      offset += valueLen;
+    }
+
+    return result;
+  }
+
+  private byte[] serializeValues(final Map<String, ByteIterator> values) throws IOException {
+    try(final ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
+      final ByteBuffer buf = ByteBuffer.allocate(4);
+
+      for(final Map.Entry<String, ByteIterator> value : values.entrySet()) {
+        final byte[] keyBytes = value.getKey().getBytes(UTF_8);
+        final byte[] valueBytes = value.getValue().toArray();
+
+        buf.putInt(keyBytes.length);
+        baos.write(buf.array());
+        baos.write(keyBytes);
+
+        buf.clear();
+
+        buf.putInt(valueBytes.length);
+        baos.write(buf.array());
+        baos.write(valueBytes);
+
+        buf.clear();
+      }
+      return baos.toByteArray();
+    }
+  }
+
+  private void createColumnFamily(final String name) throws RocksDBException {
+    COLUMN_FAMILY_LOCKS.putIfAbsent(name, new ReentrantLock());
+
+    final Lock l = COLUMN_FAMILY_LOCKS.get(name);
+    l.lock();
+    try {
+      if(!COLUMN_FAMILIES.containsKey(name)) {
+        final ColumnFamilyOptions cfOptions = new ColumnFamilyOptions().optimizeLevelStyleCompaction();
+        final ColumnFamilyHandle cfHandle = rocksDb.createColumnFamily(
```

Comment:
I'm trying to work through the reasoning around locking. is this call to `rocksDb.createColumnFamily` the non-threadsafe thing we're using the column family lock to guard?

Reply:
There are two different locking mechanisms here at play:

1. It seemed to me that there could be multiple `RocksDBClient` object instances within a running instance of YCSB. However, as RocksDB itself is an embedded database, we can only open and close it once for the entire lifecycle of YCSB. The synchronization on `RocksDBClient.class` ensures that RocksDB will only be opened once on the first call, and then only closed on the last call (ensured via reference counting).

2. A YCSB "table" is represented in RocksDB as a Column Family. We create these column families on demand, and we must only create them once (on their first use). The `COLUMN_FAMILY_LOCKS` are used to ensure that a column family is only created once no matter how many concurrent calls there are for an operation on the "table". Also note that, when you create a column family in RocksDB, the next time you open the database you have to tell it of all existing column families, otherwise it will error. This is why we also persist the column family names to a text file.

Hope that makes sense?